kernel: Add k_delayed_work_pending to check if work has been submitted

joerchan opened the pull request:
Related to my comment here:
https://github.com/zephyrproject-rtos/zephyr/issues/22803#issuecomment-585887641

I think this utility function would complement the API and would have been possible to use to handle #22756, since the cancel return value was not working as expected.

Would love some feedback if this is useful or correct.

Add k_delayed_work_pending similar to k_work_pending to check if the
delayed work item has been submitted but not yet completed.
This would compliment the API since using k_work_pending or
k_delayed_work_remaining_get is not enough to check this condition.
This is because the timeout could have run out, but the timeout handler
not yet processed and put the work into the workqueue.

joerchan committed:
kernel: Add k_delayed_work_pending to check if work has been submitted

Add k_delayed_work_pending similar to k_work_pending to check if the
delayed work item has been submitted but not yet completed.
This would compliment the API since using k_work_pending or
k_delayed_work_remaining_get is not enough to check this condition.
This is because the timeout could have run out, but the timeout handler
not yet processed and put the work into the workqueue.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
CC: @pabigot  @Vudentz

pabigot commented on include/kernel.h:
```diff
@@ -3002,6 +3002,22 @@ static inline bool k_work_pending(struct k_work *work)
+ *
+ * @param      work  Address of delayed work item.
+ *
+ * @return     true if work item is pending, or fales if it is not pending.
```
s/fales/false/

Also don't indent the formal parameter name after `@param`, this header doesn't use that style.

There's some mis-use of the term "pending", since the verbose description distinguishes "pending in a queue" from "delayed for a timeout".  I think that's already present in the existing delayed work descriptions, but it's still worth changing this to something like "true if work item is waiting for the delay to expire or pending on a work queue".

pabigot commented on include/kernel.h:
```diff
@@ -3002,6 +3002,22 @@ static inline bool k_work_pending(struct k_work *work)
+static inline bool k_delayed_work_pending(struct k_delayed_work *work)
+{
+	return !z_is_inactive_timeout(work->timeout) ||
+	       k_work_pending(work->work);
```
`&work->work`?

joerchan replied:
Thanks. I didn't actually test this. Maybe I should have submitted it as just a draft first.

I was mainly interested to know if the two functions are enough to cover the whole range from submitted to finished executing.

pabigot reviewed (dismissed):
Will approve, but there should be a test of this somewhere.

joerchan commented on include/kernel.h:
```diff
@@ -3002,6 +3002,23 @@ static inline bool k_work_pending(struct k_work *work)
+ */
+static inline bool k_delayed_work_pending(struct k_delayed_work *work)
+{
+	return !z_is_inactive_timeout(&work->timeout) ||
```
Will have to get the definition for this function here.

Vudentz replied:
Id leave a comment about this being racy, also aren't we missing the check for work being detached (work->work_q == NULL) so it matches the behavior of work_cancel.

joerchan replied:
Can add the comment.

Checking for work->worq_q == NULL I wouldn't do, as the value of work_q is not reset after the work is finished, which is a problem with work_cancel.

andyross reviewed (approved):
No complaints with this function.  It's worthwhile to have for symmetry.  But note that the API (and `k_work_pending()`) is inherently racy, and given my preference I'd say neither should exist.  Knowing that the work is pending now says nothing about whether it still will be in the next instruction where you use that knowledge.

Using this information safely requires knowing that the work timer won't fire, e.g. that you're in an ISR at a higher priority than the timer interrupt, that you hold an irq lock on a uniprocessor system (on SMP that doesn't work), that the work queue is blocked (e.g. that you're inside a callback for the same queue right now), etc...  Not really a great tool.

andrewboie commented:
the function is trivial but @pabigot is right, this must have a test.

joerchan committed:
kernel: Add note regarding safe use of the k_work_pending return value

Add note regarding safe use of the k_work_pending return value. Similar
to the note for k_delayed_work_pending.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

pabigot reviewed (commented):
As demonstrated in #23132 although `k_delayed_work_cancel()` is documented to return `-EALREADY` if the work item has been *completed*, it actually returns it if the work item has already been submitted.

I'm skeptical of the whole k_work implementation.  I'm seeing at least nine distinct states for a work item that is not cancelled:
* unsubmitted
* delayed on timeout
* timeout expired, not yet added to queue
* marked pending but not on queue
* on queue
* removed from queue
* cleared pending
* invoking handler
* complete

Most of these can't be inferred by inspecting the state.

If consensus is this is useful I won't block it, but I would prefer to see somebody take the time to go through the whole k_work API and implementation and revise it so the documentation accurately describes the behavior and the functionality is useful.  By "useful" I mean that after an attempt to cancel you should always be able to tell whether the item was:

* submitted but not yet committed to processing (delayed or on queue): can be cancelled, 0
* processing (including committed to processing, not yet in handler): cannot be cancelled, now `-EINVAL`  preferably `-EINPROGRESS`
* complete (which may be indistinguishable from unsubmitted): too late to cancel, now `-EALREADY` preferably `-EINVAL`

This capability should be present whether the item is a standard work item, a delayed work item, or a pollable work item.

joerchan commented:
@andrewboie @pabigot Anything that I should do about this PR? It's been marked as stale now.

tyrantlizard77 reviewed (approved):
> If consensus is this is useful I won't block it

I think @andyross and I are on the same page wrt fundamental raciness of how this can be used. You simply can't make decisions based on what the API tells you, if incorrect information would cause something bad to happen.

> I would prefer to see somebody take the time to go through the whole k_work API and implementation and revise it so the documentation accurately describes the behavior and the functionality is useful

@pabigot I concur. The existing API has been carried forward from the Viper days for a long time; it is workable for a lot of people but a ground-up redesign would be nice. Would you be willing to file an enhancement ticket with the design goals as you see them?

> @andrewboie @pabigot Anything that I should do about this PR? It's been marked as stale now.

Even though I'm not thrilled for reasons above, It doesn't hurt anything and we already had the same problems with an existing API. There is a test. There is a disclaimer that things can change at any time.

andrewboie commented:
@pabigot by the way this isn't the only IPC API that needs refactoring. Pretty much all the IPC objects in the kernel are *old*. I ran into all sorts of problems converting some of them to be safe from userspace (poll and queues in particular), and @andyross and I have talked about how some of them (mailboxes, etc) really should be moved out of the core kernel and into library code, built on either on top of k_sem, k_mutex, or a wait/wake abstraction using spinlocks for synchronization so they can be used from ISRs.

I at least haven't had the time to dive into redesigning them, although the direction I want to go with userspace is all futex-based, and Andy recently overhauled our heap code. I believe the core k_sem and k_mutex objects are a sound foundation for someone to build whatever IPC they like, but still...if I could clone myself this area would be near the top of my list to look into.

pabigot commented:
@joerchan OK, stale removed, you have two approvals from kernel gurus.   It's four months out of date, though, so please rebase on current master and push a new version so we can verify it's still all good.  Once that's done I'll remove DNM and somebody can merge it.

DNM until a fresh CI check has run.

pabigot reviewed (commented):
I'd rather this wasn't necessary and don't want to approve because I'm just skeptical of the entire k_work infrastructure.  But I won't block it, so it's good to merge with existing approvals.

I plan to start working #27356 very soon, so hopefully we'll have a more robust solution in this or the next release.
